client/components/edit/Breadcrumbs.tsx

```diff
@@ -1,130 +1,36 @@
+import { TreeData, generateBreadcrumbTrails, processBreadcrumbData } from "../../util/Breadcrumbs";
+import { useEditorContext } from "../../context/EditorContext";
 import Link from "next/link";
 
-interface TreeNode {
-    pid: string;
-    title: string;
-    parents: Array<TreeNode>;
-}
-
-interface TreeData {
-    topNodes: Array<string>;
-    records: Record<string, TreeNode>;
-    childLookups: Record<string, Array<string>>;
-}
-
-interface BreadcrumbTrail {
-    pid: string;
-    path: Array<TreeNode>;
-}
-
 interface BreadcrumbsProps {
     pid: string;
 }
 
 const Breadcrumbs = ({ pid = "" }: BreadcrumbsProps): React.ReactElement => {
     const {
-        action: { fetchJSON },
-    } = useFetchContext();
-    const [treeData, setTreeData] = useState<TreeData>({
-        topNodes: [],
-        childLookups: {},
-        records: {},
-    });
-
-    /**
-     * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
-     * no parents) and creating lookup tables for children. This makes it easier to
-     * render breadcrumbs from left to right and to find all relevant trails.
-     */
-    function processBreadcrumbData(data: TreeNode): TreeData {
-        const queue: Array<TreeNode> = [data];
-        const topNodes: Array<string> = [];
-        const childLookups: Record<string, Array<string>> = {};
-        const records: Record<string, TreeNode> = {};
-        while (queue.length > 0) {
-            const current = queue.shift();
-            if (!current) {
-                break;
-            }
-            if (current.parents.length === 0) {
-                topNodes.push(current.pid);
-            }
-            records[current.pid] = { pid: current.pid, title: current.title, parents: [] };
-            current.parents.forEach((parent) => {
-                queue.push(parent);
-                if (typeof childLookups[parent.pid] === "undefined") {
-                    childLookups[parent.pid] = [current.pid];
-                } else {
-                    childLookups[parent.pid].push(current.pid);
-                }
-            });
-        }
-        return {
-            topNodes: Array.from(new Set(topNodes)), // deduplicate
-            records,
-            childLookups,
-        };
-    }
+        state: { parentDetailsStorage },
+        action: { loadParentDetailsIntoStorage },
+    } = useEditorContext();
+    const loaded = Object.prototype.hasOwnProperty.call(parentDetailsStorage, pid);
 
     useEffect(() => {
-        async function loadData() {
-            let data: TreeData = {
-                topNodes: [],
-                childLookups: {},
-                records: {},
-            };
-            const url = getObjectParentsUrl(pid);
-            try {
-                data = processBreadcrumbData(await fetchJSON(url));
-            } catch (e) {
-                console.error("Problem fetching breadcrumb data from " + url);
-            }
-            setTreeData(data);
```

Comment:
This is a pretty good refactor. We usually want to abstract out the transformation logic and just leave the render logic.